### DISCUSSION TITLE
Drawing a PlotItem in different QGraphicsViews

### DISCUSSION THREAD

**Discussion by sntaountakis** [pyqtgraph/pyqtgraph#2982] [Q&A]
Hi, I have an application that, in simple terms, has some PlotItems arranged in a GraphicsLayout. I would like to transition to a logic where my PlotItems are arranged with Docks and a DockArea, but for my application it is preferable to keep the PlotItems in a layout in the same GraphicsScene.

That means, I do not want to isolate each item in its own scene and have a GraphicsView centered to it (like PlotWidget) to place it in Dock. I want to create new Views for my existing items.

I had the idea to fit each item in a QGraphicsView and overwrite the paint methods of the items to paint differently depending on the view that draws them. In theory, the Axis bounds for example, could be painted to the respective view's available space, and similarly the ViewBox etc. The actual item's geometry would stay intact (pyqtgraph.GraphicsView changes the items geometry when resizeEvent is triggered)

I wanted to ask about the feasibility of such task. Looking through the code, I understand that the geometry of the items is tightly connected with their painting operations so I do not expect this to be an easy task for sure.

Thank you!

**Comment by j9ac9k** [comment]
So you're wanting something like GraphicsLayout, where there is a shared GraphicsScene, but instead of a fixed layout, you want to be able to manipulate the position/size of the plots as if they are docks in a dock area?

I'm not sure how to go about that exactly; but I imagine it would be a large undertaking.  I have only used docks myself a little bit, but my first inclination would be to try and use docks and work around whatever limitation you have that is preventing you from doing that.

Sorry this isn't a more definitive answer, if you make headway in this area I would certainly be interesting in hearing about your progress.

**Comment by sntaountakis** [comment]
Well yes, sort of. The end goal is, as you said, to have Docks in a dock area where a user can freely resize/move each dock.

Now, each dock has a QGraphicsView widget that "looks" at a PlotItem.

The main thing here is that the item actually lives in a GraphicsLayout with other items. It has its own specific geometry and space in the scene.

What I try to take advantage of, is the fact that a QGraphicsView calls paint() on items. Therefore, instead of resizing a QGraphicsView and changing the geometry of an item (this is how pyqtgraph.GraphicsView does it), I want to paint an item within the current rectangle of the QGraphicsView that draws it.

This seemed fascinating to me as it would allow to change the way an item is painted at each view without affecting the item's geometry or the perception of the item from other views!

I hope I made it clearer. Sorry for any confusion.

**Reply by j9ac9k** [reply]
Discussing plot/gui construction on text media is hard, not your fault my imagination sucks :) 

So, you have a series of PlotItems in a GraphicsLayout, you also have a DockArea, where each Dock "references" a plot within the GraphicsLayout, which has a "fixed" geometry there.  You are looking to avoid calls to `PlotItem.resize()` when the Dock changes area, and instead want to redraw the plot from the GraphicsLayout inside the dock.  You would likely need to scale the plot to account for mismatched aspect ratios and such, but since paint method uses a `QTransform()` it's probably straight forward to account for scaling here (but you won't get new detail if I understand the situation correctly). 

The first thing that comes to mind is the pixel vector calculation in GraphicsItem: https://github.com/pyqtgraph/pyqtgraph/blob/e0b565ff95ca750b05208433ba65cb466a1dc826/pyqtgraph/graphicsItems/GraphicsItem.py#L196

This method does a lot of math to try and figure out how "big" a pixel is in the view space, attempting to "reproject" a plot to a view with a different aspect ratio (or even different size) will likely cause problems somewhere along the way.  It's a pretty low level method that is used in many places, so trying to work around it will be tough to say the least.

**Reply by sntaountakis** [reply]
Yes, you are correct this is exactly what I want to do. It is true that I could just simply apply a QTransform to scale the items but that would stretch them and lose desired details. I will start loking into pixelVectors and see how that could help me. Thank you for your help!
